### PR TITLE
Totalsegmentator dataset

### PR DESCRIPTION
**Pull request by shevtsovalexey**

### CODE REVIEW

File: amid/totalsegmentator/anatomical_structures.json [hunk 1]
```diff
@@ -0,0 +1 @@
...t_8", "rib_left_10", "vertebrae_C4", "trachea", "rib_left_1", "iliopsoas_left", "autochthon_left", "duodenum", "portal_vein_and_splenic_vein", "vertebrae_T4", "femur_right", "rib_right_9", "vertebrae_T9", "rib_right_2", "vertebrae_T5", "vertebrae_L4", "iliac_artery_left", "gluteus_maximus_left", "inferior_vena_cava", "iliopsoas_right", "kidney_right", "heart_ventricle_right", "spleen", "vertebrae_L1", "vertebrae_L3", "vertebrae_T2", "lung_upper_lobe_right", "rib_right_5", "adrenal_gland_left", "rib_right_11", "heart_atrium_left", "liver", "iliac_vena_left", "vertebrae_T12", "vertebrae_L2", "rib_right_1", "rib_left_6", "lung_lower_lobe_right", "vertebrae_T6", "small_bowel", "gluteus_minimus_left", "urinary_bladder", "rib_right_12", "rib_right_6", "vertebrae_C3", "vertebrae_L5", "heart_ventricle_left", "rib_left_4", "vertebrae_T3", "lung_lower_lobe_left", "adrenal_gland_right", "heart_myocardium", "gluteus_medius_right", "lung_middle_lobe_right", "rib_left_3", "heart_atrium_right", "vertebrae_T10", "vertebrae_C6", "humerus_right", "gluteus_maximus_right", "vertebrae_C7", "vertebrae_C5", "rib_right_3", "clavicula_left", "scapula_right", "pancreas", "vertebrae_T11", "iliac_vena_right"]
```
Comment (maxme1) [inline amid/totalsegmentator/anatomical_structures.json:1]: better create a `.py` file for that, and import the constant from it

File: amid/totalsegmentator/dataset.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+        def make_loader(anatomical_structure):
+            def loader(i, _root: Silent):
+                nii = Path(_root) / i / 'segmentations' / f'{anatomical_structure}.nii.gz'
+                return np.asarray(nibabel.load(nii).dataobj)
+
+            return loader
+
+        for anatomical_structure in anatomical_structures:
+            scope[anatomical_structure] = make_loader(anatomical_structure)
```
Comment (maxme1) [inline amid/totalsegmentator/dataset.py:75]: move this function outside, or make it a staticmethod

**Review by maxme1 — CHANGES_REQUESTED** [review]

File: amid/totalsegmentator/dataset.py [hunk 3]
```diff
@@ -60,41 +63,80 @@ class Totalsegmentator(Source):
...
+
+
+@contextlib.contextmanager
+def unpack(root: str, relative: str):
+    unpacked = Path(root) / relative
+    if unpacked.exists():
+        yield unpacked, True
+    else:
+        with zipfile.Path(root, relative).open('rb') as unpacked:
+            yield unpacked, False
```
Comment (maxme1) [inline amid/totalsegmentator/dataset.py:142]: why not just import from bimcv?

File: amid/totalsegmentator/dataset.py [hunk 4]
```diff
@@ -60,41 +63,80 @@ class Totalsegmentator(Source):
...
 
-        return sorted(result)
+            return sorted(ids)
 
     def meta(_root: Silent):
-        return pd.read_csv(Path(_root) / 'meta.csv', sep=';')
+        file = 'Totalsegmentator_dataset/meta.csv'
+
+        with ZipFile(_root) as zf:
+            return pd.read_csv(zf.open(file), sep=';').head()
```
Comment (maxme1) [inline amid/totalsegmentator/dataset.py:105]: also use unpack

File: amid/totalsegmentator/dataset.py [hunk 5]
```diff
@@ -60,41 +63,80 @@ class Totalsegmentator(Source):
...
 
-    _add_masks(locals())
+    add_masks(locals())
 
     @meta
     def ids(_root: Silent):
-        if _root is None:
-            raise ValueError('Please pass the location of the unpacked downloaded archive.')
+        with ZipFile(_root) as zf:
+            namelist = [x.rstrip('/') for x in zf.namelist()]
```
Comment (maxme1) [inline amid/totalsegmentator/dataset.py:92]: use unpack

**Review by maxme1 — CHANGES_REQUESTED** [review]

File: amid/totalsegmentator/dataset.py (hunk 3 above)
Comment (shevtsovalexey) [reply amid/totalsegmentator/dataset.py:142]: works with `.tar.gz` only

**Review by maxme1 — APPROVED** [review]

**Pull request merged**